documentation/ImplementingASemanticCheck.md

```diff
@@ -0,0 +1,813 @@
+nodes.  I would look at each of these nodes to determine the `INTENT` of
+the associated dummy argument.
+
+This combination of the traversal framework and `dummyIntent()` would give
+me the `INTENT` of all of the dummy arguments in a FUNCTION call.  Thus, I
+would have the second piece of information I needed.
+
+## Determining if the actual argument is a variable
+I also guessed that I could determine if the `evaluate::ActualArgument`
+consisted of a variable.  
+
+Once I had a symbol for the variable, I could call one of the functions:
+```C++
+  void WarnDoVarRedefine(const parser::CharBlock &, const Symbol &);
+  void CheckDoVarRedefine(const parser::CharBlock &, const Symbol &);
+```
+to emit the messages.
+
+If my plans worked out, this would give me the three pieces of information I
+needed -- the source location of the erroneous text, the `INTENT` of the dummy
+argument, and a symbol that I could use to determine whether the actual
+argument was an active DO variable.
+
+# Implementation
+
+## Adding a parse tree visitor
+I started my implementation by adding a visitor for `parser::Expr` nodes.
+Since this analysis is part of DO construct checking, I did this in
+.../lib/semantics/check-do.[cc,h].  I added a print statement to the visitor to
+verify that my new code was actually getting executed.  
+
+In check-do.h, I added the declaration for the visitor:
+
+```C++
+  void Leave(const parser::Expr &);
+```
+
+In check-do.cc, I added an (almost empty) implementation:
+
+```C++
+  void DoChecker::Leave(const parser::Expr &) {
+    std::cout << "In Leave for parser::Expr\n";
+  }
+```
+
+I then built the compiler with these changes and ran it on my test program.
+This time, I made sure to invoke semantic checking.  Here's the command I used:
+```bash
+  ./bin/f18 -fdebug-resolve-names -fdebug-dump-parse-tree -funparse-with-symbols testfun.f90
+```
+
+This produced the output:
+
+```
+  In Leave for parser::Expr
+  In Leave for parser::Expr
+  In Leave for parser::Expr
+  In Leave for parser::Expr
+  In Leave for parser::Expr
+```
+
+This made sense since the parse tree contained five `parser::Expr` nodes.
+So far, so good.  Note that a `parse::Expr` node has a field with the
+source position of the associated expression (`CharBlock source`).  So I
+now had one of the three pieces of information needed to detect and report
+errors.
+
+## Collecting the actual arguments
+To get the `INTENT` of the dummy arguments and the `semantics::Symbol` associated with the
+actual argument, I needed to find all of the actual arguments embedded in an
+expression that contained a FUNCTION call.  So my next step was to write the
+framework to walk the `evaluate::Expr` to gather all of the
+`evaluate::ActualArgument` nodes.  The code that I planned to model it on
+was the existing infrastructure that collected all of the `semantics::Symbol` nodes from an
+`evaluate::Expr`.  I found this implementation in
+.../lib/evaluate/tools.cc:
```

Comment:
I strongly recommend getting rid of all the `.../`.

Reply:
Will do.